### ISSUE TITLE
Forms Picker is setting SelectedItem=null after switching ItemSource

### ISSUE THREAD

**Issue by rachelmoor** [xamarin/Xamarin.Forms#1879]
### Description
This bug https://bugzilla.xamarin.com/show_bug.cgi?id=57912 was not moved to GitHub.

My ViewModel is bound to a Xamarin Forms picker. When my Model updates the ViewModel with a new ItemSource and SelectedItem, the ItemSource is correctly set and the SelectedItem is set initially, and then there is an erroneous call setting the SelectedItem to null. Switching the order of the XAML did not fix this issue. My ItemSource is a List of type string.

This occurs on WPF and Android emulator. I did not check on IOS

picker binding:

              <Picker x:Name="ItemPickList"
                      BindingContext="{Binding Path=TestViewModel}"
                      ItemsSource="{Binding DisplayStrings}"
                      SelectedItem="{Binding SelectedItem}"/>

Example: 

Picker Bound to VM with Data Item 1
       ItemSource = "a", "b", "c"    SelectedItem = "a"

switch Picker Binding to Data Item 2
     ItemSouce = "red", "orange", "yellow"     SelectedItem ="yellow"

switch Picker Binding back to Data Item 1
EXPECTED: ItemSource = "a", "b", "c"    SelectedItem = "a"
ACTUAL: ItemSource = "a", "b", "c"    SelectedItem = null

### Steps to Reproduce

1.  Bind ViewModel to Xamarin.Forms Picker 
2.  Pick any SelectedItem
3.  Switch ItemSource (Model) and selected new SelectedItem
4.  Switch back to previous ItemSource (Model)
5.  Selected Item will now be cleared in ViewModel

### Expected Behavior

The Model's SelectedItem will update the Picker's SelectedItem after switching ItemSource. Switching one picker between multiple ItemSources will not clear corresponding ItemSources SelectedValue.

### Actual Behavior

The SelectedItem in the Model is cleared upon switch back to a previous Model(item source)

### Basic Information

- Version with issue: 2.5.0.122203
- Last known good version: none
- IDE: Visual Studio 2017
- Platform Target Frameworks:
  - iOS:  N/A
  - Android: 7.1
  - UWP:  15063
- Affected Devices: Android, UWP, not tested on IOS

### Screenshots

![image](https://user-images.githubusercontent.com/13911774/36183150-ea646e7a-10e1-11e8-9431-81ca81c2b10e.png)

In the screenshot you can see in the emulator the selected item (first dropdown) is empty. In the Debug window you can see the new SelectedItem bound to the Picker is null. The SelectedItem of the Data is set to "b". If you look at the code, the only place where the ViewModel's SelectedItem is set is in the DataItem which is loaded before the ViewModel in the constructor, but the ViewModel's SelectedItem is still set to null. 

### Reproduction Link

C# Code:
https://github.com/rachelmoor/XamarinFormsPickerBug/blob/master/TestPickerBug/TestPickerBug/TestPickerBug/ViewModels/AboutViewModel.cs

XAML Page:
https://github.com/rachelmoor/XamarinFormsPickerBug/blob/master/TestPickerBug/TestPickerBug/TestPickerBug/Views/AboutPage.xaml

**Comment by rachelmoor** [comment]
@samhouts @rmarinho Why was this removed?

**Comment by jassmith** [comment]
Its already in the enhancements board :)

**Comment by rachelmoor** [comment]
I'm sorry but I'm having issues finding it! Would you mind linking it?

![image](https://user-images.githubusercontent.com/13911774/36699703-56b805b2-1b02-11e8-929e-3f65e54f059d.png)

**Comment by jassmith** [comment]
Its awaiting triage in the board :)

**Comment by josesimas** [comment]
Hello, 

Would anyone be able to give an update on when this will be fixed? It is quite a critical issue for us and I would appreciate any detail on when the fix will be available. 

Thanks

**Comment by BradChase2011** [comment]
@josesimas I have fixed this issue along with numerous others with the Picker, I just havent had time to cleanup the code and make a PR.  After we get out our release I will make a PR.

**Comment by josesimas** [comment]
Thanks Brad, looking forward to it.

**Comment by marksarcade** [comment]
Is this one still planning to be worked on? Just hit the issue myself thinking i had done something wrong, but reading this I realise i'm hitting the same bug.

**Comment by peteat** [comment]
I think I've got the same, or a similar issue, although I never change the `ItemsSource` (after setting in in my view model's constructor). I've got a property `SelectedInspector` (of type `InternalInspector`) on my view model which is bound to the `SelectedItem` field of the picker. Every time I pick an option in the picker it first sets the `SelectedInspector` to the selected `InternalInspector` object (as expected) but then immediately after is sets the `SelectedInspector` to _null_ :(

Are there any workarounds? Besides not using a **<Picker>**...?

**Comment by greektreat** [comment]
I am running into the same issue. I know the root cause. Not sure if's because of bad code on my side not following the standards.
WHEN you reset the Itemsource. The pickers goes through a process of resetting it self. I see this on the call stack. 
I see on the call stack a call function Xamarin.Forms.Picker.ResetItems , which triggers it to set the SelectedItem == null. But this happens AFTER  I set the source AND the I SET the SelectedItemPropery. It Looks like it is going out of sequence.

**Comment by kmcclive** [comment]
I am having the same problem as @greektreat. I change the ItemsSource and update SelectedItem to a valid value, but it later gets reset to null. Is there at least a workaround for this? I am surprised by how long this issue has been open.

**Comment by Nico04** [comment]
I personnaly found a workaround, ItemsSource must be set BEFORE SelectedItem in order to work properly

**Comment by kmcclive** [comment]
@Nico04 Thanks for the suggestion.  As far as I can tell, I am doing that both in the view model and in the XAML, but the problem persists.

**Comment by kwacks** [comment]
Yeah most of these workarounds don't work, the value still gets set to null. I ended up setting SelectedItem in OnBindingContextChanged in the meantime.

**Comment by rachelmoor** [comment]
@kwacks @kmcclive try using the picker mentioned here: 

https://github.com/Oceanware/XamarinFormsBindablePicker/blob/master/XamarinFormsBindablePicker/XamarinFormsBindablePicker/Controls/BindablePicker.cs

This solved the problem for me!

**Comment by kmcclive** [comment]
@rachelmoor Thanks for the suggestion.  I'll give that a try!

**Comment by Josecanalla** [comment]
@rachelmoor it works for me, thanks!

In my case the Picker control works OK in iOS but in Android I get the SelectedItem null.

**Comment by samhouts** [comment]
This issue doesn't seem to have had any activity in a long time. We're working on prioritizing issues and resolving them as quickly as we can. To help us get through the list, we would appreciate an update from you to let us know if this is still affecting you on the latest version of Xamarin.Forms, since it's possible that we may have resolved this as part of another related or duplicate issue. If we don't see any new activity on this issue in the next 30 days, we'll evaluate whether this issue should be closed. Thank you!

**Comment by kmcclive** [comment]
I'm still seeing some similar behavior, but don't know what is expected.  If SelectedItem is set and I change the ItemsSource to a collection that still contains the SelectedItem, SelectedItem is then set to null.  Is this expected?  Do I need to explicitly set SelectedItem whenever I change the ItemsSource?

**Comment by EmmanuelJego** [comment]
We used to be affected by this issue, but after upgrading to Xamarin.Forms 5.x it seems to work like a charm! 👌

**Comment by spelltwister** [comment]
This is still an issue in 4.8.0.1821.  Now that I see 5 is out of -pre, I'll have to upgrade and test.

**Comment by ahoke-cr** [comment]
I am still experiencing this issue on Xamarin.Forms 5.0.0.1874

**Comment by iBeta666** [comment]
Me too, on 5.0.0.2012. Beyond annoying.

**Comment by YiannisBourkelis** [comment]
I run into this same issue and took me hours to understand what was going on until I found that other users are experiencing the same problem. 

Proposed behaviour to correct this issue:
The SelectedItem should not be set to null when we set a new ItemSource that contains the object referenced by the SelectedItem.
